pkg/controller/operators/catalog/step.go

```diff
@@ -0,0 +1,160 @@
+package catalog
+
+import (
+	"fmt"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/v1alpha1"
+	crdlib "github.com/operator-framework/operator-lifecycle-manager/pkg/lib/crd"
+	index "github.com/operator-framework/operator-lifecycle-manager/pkg/lib/index"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/lib/operatorclient"
+	errorwrap "github.com/pkg/errors"
+	logger "github.com/sirupsen/logrus"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Stepper manages cluster interactions based on the step.
+type Stepper interface {
+	Status() (v1alpha1.StepStatus, error)
+}
+
+// StepperFunc fulfills the Stepper interface.
+type StepperFunc func() (v1alpha1.StepStatus, error)
+
+func (s StepperFunc) Status() (v1alpha1.StepStatus, error) {
+	return s()
+}
+
+// Builder holds clients and data structures required for the StepBuilder to work
+// Builder attributes are not to meant to be accessed outside the StepBuilder method
+type builder struct {
+	opclient          operatorclient.ClientInterface
+	dynamicClient     dynamic.Interface
+	csvToProvidedAPIs map[string]cache.Indexer
+}
+
+func newBuilder(opclient operatorclient.ClientInterface, dynamicClient dynamic.Interface, csvToProvidedAPIs map[string]cache.Indexer) *builder {
+	return &builder{
+		opclient:          opclient,
+		dynamicClient:     dynamicClient,
+		csvToProvidedAPIs: csvToProvidedAPIs,
+	}
+}
+
+type notSupportedStepperErr struct {
+	message string
+}
+
+func (n notSupportedStepperErr) Error() string {
+	return n.message
+}
+
+// step is a factory that creates StepperFuncs based on the Kind provided and the install plan step.
+func (b *builder) step(stepStatus v1alpha1.StepStatus, resource v1alpha1.StepResource, name string) (Stepper, error) {
```

Comment:
It seems like `builder` is really the `Stepper` because it's -- indirectly -- performing the step. It also needs to have knowledge of all the `Stepper` implementations.

What if:
- A `Stepper` has a `Step(take *v1alpha1.Step) (taken bool, err error)` method, which performs an in-place update of the given step -- returning `taken=true` if the Stepper cared to take a step.
- Different `Stepper` implementations are made for specific resource types; e.g. `CSV`, `CRD`, `CR`, etc...
- A `StepperChain` `Stepper` implementation is made that calls a set of steppers in order until `taken=true` or an error is encountered -- forwarding the results along, or if no step was taken, returns `taken=false`.

I think something similar to this will let us view the stepper order and inclusion more easily; e.g.:
```go
// On the Operator struct
o.stepper = StepperChain{
    NewCRDStepper(...),
    NewCSVStepper(...),
    NewDefaultStepper(...),
}
// ...
// In the InstallPlan transition
taken, err := o.stepper.Step(take)
if err != nil {
    // ...
}
if !taken {
    // ...
}
// ...
```

Reply:
or maybe a `continue bool` return would be more flexible than `taken` -- the semantics would follow the namesake

Reply:
I'm down to think about this more next week